File: paperdb/src/androidTest/java/io/paperdb/DataTest.java
package io.paperdb;

import android.support.test.runner.AndroidJUnit4;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.GregorianCalendar;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import io.paperdb.testdata.Person;
import io.paperdb.testdata.PersonArg;

import static android.support.test.InstrumentationRegistry.getTargetContext;
import static io.paperdb.testdata.TestDataGenerator.genPerson;
import static io.paperdb.testdata.TestDataGenerator.genPersonList;
import static io.paperdb.testdata.TestDataGenerator.genPersonMap;
import static org.assertj.core.api.Assertions.assertThat;

/**
 * Tests List write/read API
 */
@RunWith(AndroidJUnit4.class)
public class DataTest {

    @Before
    public void setUp() throws Exception {
        Paper.INSTANCE.init(getTargetContext().getFilesDir());
        Paper.INSTANCE.book().destroy();
    }

    @Test
    public void testPutEmptyList() throws Exception {
        final List<Person> inserted = genPersonList(0);
        Paper.INSTANCE.book().write("persons", inserted);
        assertThat(Paper.INSTANCE.book().<List>read("persons")).isEmpty();
    }

    @Test
    public void testPutGetList() {
        final List<Person> inserted = genPersonList(10000);
        Paper.INSTANCE.book().write("persons", inserted);
        List<Person> persons = Paper.INSTANCE.book().read("persons");
        assertThat(persons).isEqualTo(inserted);
    }

    @Test
    public void testPutMap() {
        final Map<Integer, Person> inserted = genPersonMap(10000);
        Paper.INSTANCE.book().write("persons", inserted);

        final Map<Integer, Person> personMap = Paper.INSTANCE.book().read("persons");
        assertThat(personMap).isEqualTo(inserted);
    }

    @Test
    public void testPutPOJO() {
        final Person person = genPerson(new Person(), 1);
        Paper.INSTANCE.book().write("profile", person);

        final Person savedPerson = Paper.INSTANCE.book().read("profile");
        assertThat(savedPerson).isEqualTo(person);
        assertThat(savedPerson).isNotSameAs(person);
    }

    @Test
    public void testPutSubAbstractListRandomAccess() {
        final List<Person> origin = genPersonList(100);
        List<Person> sublist = origin.subList(10, 30);
        testReadWriteWithoutClassCheck(sublist);
    }

    @Test
    public void testPutSubAbstractList() {
        final LinkedList<Person> origin = new LinkedList<>(genPersonList(100));
        List<Person> sublist = origin.subList(10, 30);
        testReadWriteWithoutClassCheck(sublist);
    }

    @Test
    public void testPutLinkedList() {
        final LinkedList<Person> origin = new LinkedList<>(genPersonList(100));
        testReadWrite(origin);
    }

    @Test
    public void testPutArraysAsLists() {
        testReadWrite(Arrays.asList("123", "345"));
    }

    @Test
    public void testPutCollectionsEmptyList() {
        testReadWrite(Collections.emptyList());
    }

    @Test
    public void testPutCollectionsEmptyMap() {
        testReadWrite(Collections.emptyMap());
    }

    @Test
    public void testPutCollectionsEmptySet() {
        testReadWrite(Collections.emptySet());
    }

    @Test
    public void testPutSingletonList() {
        testReadWrite(Collections.singletonList("item"));
    }

    @Test
    public void testPutSingletonSet() {
        testReadWrite(Collections.singleton("item"));
    }

    @Test
    public void testPutSingletonMap() {
        testReadWrite(Collections.singletonMap("key", "value"));
    }

    @Test
    public void testPutGeorgianCalendar() {
        testReadWrite(new GregorianCalendar());
    }

    @Test
    public void testPutSynchronizedList() {
        testReadWrite(Collections.synchronizedList(new ArrayList<>()));
    }

    @Test()
    public void testReadWriteClassWithoutNoArgConstructor() {
        testReadWrite(new PersonArg("name"));
    }

    private Object testReadWriteWithoutClassCheck(Object originObj) {
        Paper.INSTANCE.book().write("obj", originObj);
        Object readObj = Paper.INSTANCE.book().read("obj");
        assertThat(readObj).isEqualTo(originObj);
        return readObj;
    }

    private void testReadWrite(Object originObj) {
        Object readObj = testReadWriteWithoutClassCheck(originObj);
        assertThat(readObj.getClass()).isEqualTo(originObj.getClass());
    }

}

File: paperdb/src/androidTest/java/io/paperdb/DataTest.kt
package io.paperdb

import android.support.test.runner.AndroidJUnit4

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

import java.util.ArrayList
import java.util.Arrays
import java.util.Collections
import java.util.GregorianCalendar
import java.util.LinkedList

import io.paperdb.testdata.Person
import io.paperdb.testdata.PersonArg

import android.support.test.InstrumentationRegistry.getTargetContext
import io.paperdb.testdata.TestDataGenerator.genPerson
import io.paperdb.testdata.TestDataGenerator.genPersonList
import io.paperdb.testdata.TestDataGenerator.genPersonMap
import kotlinx.coroutines.runBlocking
import org.assertj.core.api.Assertions.assertThat

/**
 * Tests List write/read API
 */
@RunWith(AndroidJUnit4::class)
class DataTest {

    @Before
    @Throws(Exception::class)
    fun setUp() {
        Paper.init(getTargetContext().filesDir)
        Paper.book().destroy()
    }

    @Test
    @Throws(Exception::class)
    fun testPutEmptyList() = runBlocking<Unit> {
        val inserted = genPersonList(0)
        Paper.book().write<Any>("persons", inserted)
        assertThat(Paper.book().read<List<*>>("persons")).isEmpty()
    }

    @Test
    fun testPutGetList() = runBlocking<Unit> {
        val inserted = genPersonList(10000)
        Paper.book().write<Any>("persons", inserted)
        val persons = Paper.book().read<List<Person>>("persons")
        assertThat(persons).isEqualTo(inserted)
    }

    @Test
    fun testPutMap() = runBlocking<Unit> {
        val inserted = genPersonMap(10000)
        Paper.book().write<Any>("persons", inserted)

        val personMap = Paper.book().read<Map<Int, Person>>("persons")
        assertThat(personMap).isEqualTo(inserted)
    }

    @Test
    fun testPutPOJO() = runBlocking<Unit> {
        val person = genPerson(Person(), 1)
        Paper.book().write<Any>("profile", person)

        val savedPerson = Paper.book().read<Person>("profile")
        assertThat(savedPerson).isEqualTo(person)
        assertThat(savedPerson).isNotSameAs(person)
    }

    @Test
    fun testPutSubAbstractListRandomAccess() {
        val origin = genPersonList(100)
        val sublist = origin.subList(10, 30)
        testReadWriteWithoutClassCheck(sublist)
    }

    @Test
    fun testPutSubAbstractList() {
        val origin = LinkedList(genPersonList(100))
        val sublist = origin.subList(10, 30)
        testReadWriteWithoutClassCheck(sublist)
    }

    @Test
    fun testPutLinkedList() {
        val origin = LinkedList(genPersonList(100))
        testReadWrite(origin)
    }

    @Test
    fun testPutArraysAsLists() {
        testReadWrite(Arrays.asList("123", "345"))
    }

    @Test
    fun testPutCollectionsEmptyList() {
        testReadWrite(emptyList<Any>())
    }

    @Test
    fun testPutCollectionsEmptyMap() {
        testReadWrite(emptyMap<Any, Any>())
    }

    @Test
    fun testPutCollectionsEmptySet() {
        testReadWrite(emptySet<Any>())
    }

    @Test
    fun testPutSingletonList() {
        testReadWrite(listOf("item"))
    }

    @Test
    fun testPutSingletonSet() {
        testReadWrite(setOf("item"))
    }

    @Test
    fun testPutSingletonMap() {
        testReadWrite(Collections.singletonMap("key", "value"))
    }

    @Test
    fun testPutGeorgianCalendar() {
        testReadWrite(GregorianCalendar())
    }

    @Test
    fun testPutSynchronizedList() {
        testReadWrite(Collections.synchronizedList(ArrayList<Any>()))
    }

    @Test
    fun testReadWriteClassWithoutNoArgConstructor() {
        testReadWrite(PersonArg("name"))
    }

    private fun testReadWriteWithoutClassCheck(originObj: Any): Any? = runBlocking {
        Paper.book().write("obj", originObj)
        val readObj = Paper.book().read<Any>("obj")
        assertThat(readObj).isNotNull
        assertThat(readObj).isEqualTo(originObj)
        readObj
    }

    private fun testReadWrite(originObj: Any) {
        val readObj = testReadWriteWithoutClassCheck(originObj)
        assertThat(readObj!!.javaClass).isEqualTo(originObj.javaClass)
    }

}
